rust/hyperlane-core/src/traits/pending_operation.rs

```diff
@@ -85,11 +101,39 @@ pub trait PendingOperation: Send + Sync + Debug + TryBatchAs<HyperlaneMessage> {
     /// retried immediately.
     fn reset_attempts(&mut self);
 
-    #[cfg(test)]
     /// Set the number of times this operation has been retried.
+    #[cfg(any(test, feature = "test-utils"))]
     fn set_retries(&mut self, retries: u32);
 }
 
+/// Utility fn to calculate the total estimated cost of an operation batch
+pub fn total_estimated_cost(ops: &[Box<dyn PendingOperation>]) -> U256 {
+    ops.iter()
+        .fold(U256::zero(), |acc, op| match op.get_tx_cost_estimate() {
+            Some(cost_estimate) => acc.saturating_add(cost_estimate),
+            None => {
+                warn!(operation=?op, "No cost estimate available for operation, defaulting to 0");
+                acc
+            }
+        })
+}
+
+/// Calculate the gas used by an operation in a batch, by looking at the total cost of the batch,
+/// and the estimated cost of the operation compared to the sum of the estimates of all operations in the batch.
+pub fn gas_used_by_batch_operation(
+    batch_outcome: &TxOutcome,
+    batch_estimated_cost: U256,
+    operation_estimated_cost: U256,
+) -> ChainResult<U256> {
+    let gas_used_by_batch = FixedPointNumber::try_from(batch_outcome.gas_used)?;
+    let gas_limit = FixedPointNumber::try_from(operation_estimated_cost)?;
```

Comment:
nit: i had a bit of a hard time following this logic bc the variable names of the fixed versions didn't match the param names, might be easier to read if they stay similar